filter/lua.go

```diff
@@ -0,0 +1,342 @@
+- lua: ` + ticks("validateRecord(record) -> (bool, int)") + `
+
+` + ticks("validateRecord") + ` is the lua equivalent of the ` + ticks("ValidateRecord") + ` function passed to your filter during construction.
+It validates a given record with respect to the validation function, returning a boolean indicating whether 
+the record is a valid one, if false, the returned integer indicates the index of the first invalid field it met.
+
+#### fieldByName
+
+- Go: ` + ticks("Components.FieldByName(string) (FieldIndex, bool)") + `
+- lua: ` + ticks("fieldByName(string) -> int|nil") + `
+
+` + ticks("fieldByName") + ` is the lua equivalent of the ` + ticks("FieldByName") + ` function passed to your filter during construction.
+It allows to lookup a field index by its name, returning the index or nil if no field exists with this index.
+
+### Global tables
+
+#### fieldNames
+
+- Go: ` + ticks("Components.FieldNames []string") + `
+- lua: ` + ticks("fieldNames") + `
+
+` + ticks("fieldNames") + ` is an integed-indexed table, in other words an array, containing all field names, as ` + ticks("FieldNames") + ` in Go.
+`
+
+// TODO(arl): ideally this functions should not be required, but writing
+// markdown documentation in Go strings is tedious and error-prone.
+
+func ticks(s string) string         { return "`" + s + "`" }
+func startBlock(lang string) string { return "```" + lang }
+func endBlock() string              { return "```" }
+
+// LUADesc describes the LUA filter
+var LUADesc = baker.FilterDesc{
+	Name:   "LUA",
+	New:    NewLUA,
+	Config: &LUAConfig{},
+	Help:   luaHelp,
+}
+
+// LUAConfig holds the configuration for the LUA filter.
+type LUAConfig struct {
+	Script     string `help:"Path to the lua script where the baker filter is defined" required:"true"`
+	FilterName string `help:"Name of the lua function to run as baker filter" required:"true"`
+}
+
+// LUA allows to run a baker filter from an external script written in lua.
+type LUA struct {
+	l       *lua.LState    // lua state used during all the baker filter lifetime
+	ud      *lua.LUserData // pre-allocated (reused) userdata for the processed record
+	luaFunc lua.LValue     // lua filter function
+	luaNext *lua.LFunction // lua next function (reused)
+	next    func(baker.Record)
+}
+
+// NewLUA returns a new LUA filter.
+func NewLUA(cfg baker.FilterParams) (baker.Filter, error) {
+	dcfg := cfg.DecodedConfig.(*LUAConfig)
+
+	l := lua.NewState()
+	if err := l.DoFile(dcfg.Script); err != nil {
+		return nil, fmt.Errorf("can't compile lua script %q: %v", dcfg.Script, err)
+	}
+
+	registerLUATypes(l, cfg.ComponentParams)
+
+	luaFunc := l.GetGlobal(dcfg.FilterName)
+	if luaFunc.Type() == lua.LTNil {
+		return nil, fmt.Errorf("can't find lua filter %q in script %q", dcfg.FilterName, dcfg.Script)
+	}
+
+	// Preallocate the userdata we use to wrap the record passed to the filter.
+	// We can do this since a single instance of a baker filter is only ever
```

Comment:
You're right good catch. The filter was made to only work with a non-concurrent filter chain, which could be ok. But since the filter has no way to know the current value of the `[filterchain] procs` (and it's better in this way) I made the necessary changes to make the filter work in any situation. In other words I removed the pre-allocations and protected the lua state between concurrent calls to `Process`.

Reply:
wrt the proposal to create a new user structure / state for each record, it's probably worth measure the impact of the operation with a huge amount of records (as we usually have)